### PR TITLE
Workshop Custom Transactions - LiskBills

### PR DESCRIPTION
**Pull request by michielmulders**
Ready to use workshop for internal use and Meetup event to try out this format. How easy is it for users to read up and solve tasks / write some code themselves?

**Included:**
- Prerequisites
- Concept explanation LiskBills
- Setup
- Transaction 1: Invoice Transaction (read code and comprehend)
- Extra nuggets of knowledge on custom transactions
- Transaction 2: Payment Transaction (task is given, user has to re-create payment transaction in small steps)
- Testing the Payment Transaction

### CODE REVIEW

File: workshop/README.md [hunk 1]
```diff
@@ -1,26 +1,34 @@
...
+2. Checkout branch `development` (by default active branch).<br><br>
 
-## Setup
-1. Clone [Lisk-SDK-Examples](https://github.com/LiskHQ/lisk-sdk-examples) repository locally and checkout branch `workshop-custom-txs-start`.
+2. Navigate inside `/workshop` folder and run `npm install` to install the required dependencies for the workshop.<br><br>
+
+3. Next, the `/workshop/transactions` folder is considered a separate module which holds the code for the transaction types. Again, we have to execute `npm install` inside this folder to install the remaining dependencies. Notice in both `package.json` files, there is only one dependency we use: `lisk-sdk: 2.3.5`.<br><br>
+
+4. To verify the setup is correct, try to run the application with `npm start`. The `npm start` command will run the `index.js` file and pipe the outputted logs to our preferred log formatting tool Bunyan (wraps: `node index.js | npx bunyan -o short`). If everything is running fine, you can go to the next section.<br><br>_In case you have experimented before with Lisk, it is possible you have to first drop your database and recreate it with:_ `dropdb lisk_dev && createdb lisk_dev`.
```
Comment (Tschakki) [inline workshop/README.md]: Are we really recommending to use `npm start` with lisk sdk?

**Review by Tschakki — CHANGES_REQUESTED** [review]
- [x] https://github.com/LiskHQ/lisk-sdk-examples/blame/workshop-custom-txs-start/workshop/README.md#L56 The fee is returned as string, isn’t it?

- [x] https://github.com/LiskHQ/lisk-sdk-examples/blame/workshop-custom-txs-start/workshop/README.md#L67 here you can link directly to line 399: https://github.com/LiskHQ/lisk-sdk/blob/development/elements/lisk-transactions/src/base_transaction.ts#L399

- [x] https://github.com/LiskHQ/lisk-sdk-examples/blame/workshop-custom-txs-start/workshop/README.md#L147 I thought it is only applied from the moment it gets included in a block?

- [x] https://github.com/LiskHQ/lisk-sdk-examples/blame/workshop-custom-txs-start/workshop/README.md#L314 - I wouldn’t use “I” here, sounds weird, I think it’s the first time you use it here

- [x] https://github.com/LiskHQ/lisk-sdk-examples/blame/workshop-custom-txs-start/workshop/README.md#L345 here we should mention to uncomment the payment tx in index.js

File: workshop/README.md (hunk 1 above)
Comment (michielmulders) [reply workshop/README.md]: I'll update that, better not right?

File: workshop/README.md [hunk 2]
```diff
@@ -1,18 +1,34 @@
...
-## Concept: LiskBills
-This workshop will cover some aspects of using the Alpha SDK for developing custom transactions.
-First, we will explore two custom transactions which we used for LiskBills. The idea is that we can send an `invoice_transaction` to send an invoice to a client. Next, the client can fulfill the invoice by sending a `payment_transaction`.
+**Make sure to read the prerequisites and execute each step of the setup.**
+
+## Prerequisites
+- Code editor like _Visual Studio Code_
+- Node.js (v10 or higher installed - use [nvm](https://github.com/nvm-sh/nvm#install--update-script) for easy switching between versions)
+- Have `curl` or `Postman` installed
+- Have Postgres **system-wide** installed, **don't install via Docker**! ([Installation guide by Lisk](https://lisk.io/documentation/lisk-core/setup/source#b-postgres-system-wide)) Notice, we also talk about creating a lisk user on your system, for MacOS this is not needed.
```
Comment (Tschakki) [inline workshop/README.md]:
```suggestion
- Have Postgres installed ([Installation guide by Lisk](https://lisk.io/documentation/lisk-core/setup/source#b-postgres-system-wide)). Notice, we also talk about creating a `lisk` user on your system, for MacOS this is not needed.
```

File: workshop/README.md [hunk 3]
```diff
@@ -1,18 +1,34 @@
...
+1. Clone [Lisk-SDK-Examples](https://github.com/LiskHQ/lisk-sdk-examples) repository locally.<br><br>
+
+2. Checkout branch `development` (by default active branch).<br><br>
+
+2. Navigate inside `/workshop` folder and run `npm install` to install the required dependencies for the workshop.<br><br>
+
+3. Next, the `/workshop/transactions` folder is considered a separate module which holds the code for the transaction types. Again, we have to execute `npm install` inside this folder to install the remaining dependencies. Notice in both `package.json` files, there is only one dependency we use: `lisk-sdk: 2.3.5`.<br><br>
+
+4. To verify if the setup is correct, try to run the application with `node index.js | npx bunyan -o short` from the root of the `/workshop` folder. This command will run the `index.js` file and pipe the outputted logs to our preferred log formatting tool Bunyan. If everything is running fine, you can go to the next section. To stop the node, hit CTRL+C.<br><br>_In case you have experimented before with Lisk, it is possible you have to first drop your database and recreate it with:_ `dropdb lisk_dev && createdb lisk_dev`.
```
Comment (Tschakki) [inline workshop/README.md]:
```suggestion
4. To verify if the setup is correct, try to run the application with `node index.js | npx bunyan -o short` from the root of the `/workshop` folder. This command will run the `index.js` file and pipe the outputted logs to our preferred log formatting tool Bunyan. If everything is running fine, you can go to the next section. To stop the node, hit CTRL+C.<br><br>_In case you have experimented before with Lisk, it is possible you have to first drop your database and recreate it with:_ `dropdb lisk_dev && createdb lisk_dev --owner lisk`.
```

File: workshop/README.md [hunk 4]
```diff
@@ -1,18 +1,34 @@
...
 
-3. In order to verify the setup is correct, try to run the application with `npm start`. The `npm start` command will run the `index.js` file and pipe the outputted logs to our preferred log formatting tool Bunyan (wraps: `node index.js | npx bunyan -o short`). If everything is running fine, you can go to the next section.
+For the above reason, Bob asks Alice to create the invoice via a Lisk custom transaction (InvoiceTransaction). Next, he creates another custom transaction to send the payment to Alice (PaymentTransaction).
+
+**In summary:** We will explore the above two custom transaction types that we used for LiskBills. The idea is that a freelancer can send an `invoice_transaction` to send an invoice to a client. Next, the client can pay for the invoice by sending a `payment_transaction` to the freelancer.
 
 ## Transaction 1: Invoice Transaction
 The invoice transaction accepts three parameters (send by the freelancer to the client):
```
Comment (Tschakki) [inline workshop/README.md:35]: Maybe we should also mention the params that are required by default for every transaction, like timestamp, fee, recipientId? I think it makes sense to explain it because it can be confusing e.g. for transactions that do not need a recipient, if we still include one when creating the tx object, but not mention in the tutorial why we are including it.

File: workshop/README.md [hunk 5]
```diff
@@ -25,43 +41,47 @@ The invoice transaction accepts three parameters (send by the freelancer to the
...
 
-**Task 1: Go to [BaseTransaction class](https://github.com/LiskHQ/lisk-sdk/blob/development/elements/lisk-transactions/src/base_transaction.ts) at line 399 to copy the implementation and paste it in the prepare() function for InvoiceTransaction replacing the `super` call.**
+However, this is a bad practice as the implementation might change inside the `BaseTransaction`. Therefore, we want to open the `BaseTransaction` to copy the code that caches the sender account.<br><br>
 
-_The solution for each step can be found in the `Solution: Invoice Transaction` section._
+**Task 1: Go to [BaseTransaction class](https://github.com/LiskHQ/lisk-sdk/blob/development/elements/lisk-transactions/src/base_transaction.ts#L400-L404) and copy the highlighted implementation of the `prepare()` function (don't copy the whole function definition as it's created in TypeScript) and paste it in the `prepare()` function for InvoiceTransaction replacing the `super` call.**
```
Comment (Tschakki) [inline workshop/README.md]: The highlighted code snippet seems wrong to me. Doesn't fit to the solution as well.

File: workshop/README.md [hunk 6]
```diff
@@ -125,35 +161,45 @@ and stores this data in an in-memory key-value store inside the Lisk application
...
-_The solution can be found in the `Solution: Invoice Transaction` section._
+_The solution can be found in the `Solution: Invoice Transaction` section._<br><br>
 
 ### Testing the InvoiceTransaction
 To test the `InvoiceTransaction`, we have to register the custom transaction to our blockchain in the `index.js` file in the root of the `/workshop` folder.
 
 You'll find code that is commented out. Uncomment the line for importing the `InvoiceTransaction` and uncomment the line of code that registers the custom transaction to the application.
 
-Next, verify if everything is fine by starting the application with `npm start`. In order to verify if our custom invoice transaction works, we should send an invoice transaction. The `/workshop/generator/invoice.js` file contains a generator that uses a genesis account with sufficient funds and creates a JSON object (Run `node generator/invoice.js`).
+![Register custom transactoin](./assets/01-index-register.png)
```
Comment (Tschakki) [inline workshop/README.md:190]: Why are we using an image instead of a code snippet here?

File: workshop/README.md [hunk 7]
```diff
@@ -201,46 +277,65 @@ If you want to be absolutely sure the transaction has been accepted and included
...
+			store.account.set(sender.address, sender); // Save updated sender account
+		}
+
+		return errors;
+	}
 </details>
+<br>
 
-## Nuggets of Knowledge for Custom Transactions
+## Nuggets of Knowledge about Custom Transactions
```
Comment (Tschakki) [inline workshop/README.md:321]: Maybe not needed to fix now, but for the future:
This section is breaking a little the flow of the workshop.
As I see it, the knowledge nuggets we provide here are also needed in the second part of the workshop, when implementing the payment transaction.
To not break the flow, I would explain every "nugget" at the moment when it gets relevant in the workshop. At this point, the users might be confused how deep they need to dive into the topics to finish the workshop, or if this knowledge is needed in this context at all.

File: workshop/README.md [hunk 8]
```diff
@@ -250,17 +345,16 @@ The payment transaction accepts one parameter in the `asset.data` field (send by
...
+If you haven't installed the dependencies inside the `./transactions` folder yet, navigate in your terminal to the `./transactions` folder. Run `npm install` to again install the required dependencies as we treat the `/transactions` folder as a separate module.
 
-Now, let's get technical. Open the file at `./transactions/payment_transaction.js`. For this transaction, most code needs to be written. Good luck!
+Now, let's get technical. Open the file at `./transactions/payment_transaction.js`. For this transaction type (14), most code needs to be written. Good luck!
 
 ### Exploring Payment Transaction
-The idea for the `PaymentTransaction` is that you write the logic yourself based on the requirements and tips I give. The payment transaction will extend from the `TransferTransaction` although we do not recommend this (it simplifies things).
+The idea for the `PaymentTransaction` is that you write the logic yourself based on the requirements and tips we give to you. The payment transaction will extend from the `TransferTransaction` although we do not recommend this (it simplifies things).
```
Comment (Tschakki) [inline workshop/README.md]: We are not very consistent on this. Though we do not recommend to extend from existing tx types, we do that in Cashback and Lisk Bills to simplify things.

I think it's bad to introduce best practice and then immediatley break with it in your own tutorial. Either implement the tx type from Basetransaction or drop the comment here imho.

File: workshop/README.md [hunk 9]
```diff
@@ -270,79 +364,93 @@ The idea for this transaction is that it will accept only one value that represe
 }
 ```
 
-**Note:** For each step, the solution can be found in the section below called `Solution: Payment Transaction`.
+**Note:** For each step, the solution can be found in the section below called `Solution: Payment Transaction`.<br><br>
 
 #### Steps To Be Implemented
-1. Replace the `super` cache call with the [actual implementation in TransferTransaction](https://github.com/LiskHQ/lisk-sdk/blob/development/elements/lisk-transactions/src/0_transfer_transaction.ts#L69). Remember the task about reading up on combining AND and OR, this is exactly what happens here.<br><br>
+1. Replace the `super` cache call with the [actual implementation in the TransferTransaction class](https://github.com/LiskHQ/lisk-sdk/blob/development/elements/lisk-transactions/src/0_transfer_transaction.ts#L69).
```
Comment (Tschakki) [inline workshop/README.md:371]: Why is this needed?

**Review by Tschakki — CHANGES_REQUESTED** [review]

File: workshop/README.md (hunk 5 above)
Comment (michielmulders) [reply workshop/README.md]: You are right, weird, I've updated it to be the right selection of lines thanks for catching!

File: workshop/README.md (hunk 6 above)
Comment (michielmulders) [reply workshop/README.md:190]: Wanted to add some visuals to make the text bit more readable/segregated.

File: workshop/README.md (hunk 7 above)
Comment (michielmulders) [reply workshop/README.md:321]: Hmm yea I agree, but there are some many nuggets of knowledge that are important that I decided to add them in between. Otherwise, the user will be bombarded with reading sections. The idea is that they can implement the payment transaction on their selves uninterrupted but I see the point. Let's ask for feedback about this during the workshop to see if we need to change this format!

File: workshop/README.md (hunk 8 above)
Comment (michielmulders) [reply workshop/README.md]: Will drop the comment, indeed, need to address this to be consistent.

File: workshop/README.md (hunk 9 above)
Comment (michielmulders) [reply workshop/README.md:371]: In case the logic for transfertransaction changes, it might cache something else. But should be better removed in this case.
